### PR TITLE
cmd/snap-repair: use E-Tags when refetching a repair to retry

### PR DESCRIPTION
**Pull request by pedronis**

### CODE REVIEW

File: cmd/snap-repair/runner.go [hunk 1]
```diff
@@ -69,14 +84,17 @@ var (
...
+	ErrRepairNotModified = errors.New("repair was not modified")
+)
 
 var (
 	maxRepairScriptSize = 24 * 1024 * 1024
 )
 
-// Fetch retrieves a stream with the repair with the given ids and any auxiliary assertions.
-func (run *Runner) Fetch(brandID, repairID string) (r []asserts.Assertion, err error) {
+// Fetch retrieves a stream with the repair with the given ids and any auxiliary assertions. If revision>=0 Fetch sets in the request the If-None-Match header with an E-Tag based on it, and ErrRepairNotModified is returned if the revision is still current.
```
Comment (zyga) [inline cmd/snap-repair/runner.go]: Docstring is way too long.

File: cmd/snap-repair/runner.go [hunk 2]
```diff
@@ -186,3 +221,251 @@ func (run *Runner) Peek(brandID, repairID string) (headers map[string]interface{
...
+	Device    deviceInfo                `json:"device"`
+	Sequences map[string][]*RepairState `json:"sequences,omitempty"`
+}
+
+func (run *Runner) readState() error {
+	r, err := os.Open(dirs.SnapRepairStateFile)
+	if err != nil {
+		return err
+	}
+	dec := json.NewDecoder(r)
```
Comment (zyga) [inline cmd/snap-repair/runner.go]: missing `defer r.Close()`?

File: cmd/snap-repair/runner.go [hunk 3]
```diff
@@ -186,3 +221,251 @@ func (run *Runner) Peek(brandID, repairID string) (headers map[string]interface{
...
+	// best-effort remove old
+	os.Remove(dirs.SnapRepairStateFile)
+	// TODO: init Device
+	run.SaveState()
+	return nil
+}
+
+// LoadState loads the repairs' state from disk, and (re)initializes it if it's missing or corrupted.
+func (run *Runner) LoadState() error {
+	err := run.readState()
```
Comment (zyga) [inline cmd/snap-repair/runner.go]: Combine on one line?

File: cmd/snap-repair/runner.go [hunk 4]
```diff
@@ -186,3 +221,251 @@ func (run *Runner) Peek(brandID, repairID string) (headers map[string]interface{
...
+	}
+	p := filepath.Join(d, fmt.Sprintf("repair.r%d", r[0].Revision()))
+	return osutil.AtomicWriteFile(p, buf.Bytes(), 0600, 0)
+}
+
+func (run *Runner) readSavedStream(brandID string, seq, revision int) (r []asserts.Assertion, err error) {
+	repairID := strconv.Itoa(seq)
+	d := filepath.Join(dirs.SnapRepairAssertsDir, brandID, repairID)
+	p := filepath.Join(d, fmt.Sprintf("repair.r%d", revision))
+	f, err := os.Open(p)
```
Comment (zyga) [inline cmd/snap-repair/runner.go]: Do we need to `defer f.Close()`?

File: cmd/snap-repair/runner.go [hunk 5]
```diff
@@ -69,14 +84,20 @@ var (
...
 
 var (
 	maxRepairScriptSize = 24 * 1024 * 1024
 )
 
-// Fetch retrieves a stream with the repair with the given ids and any auxiliary assertions.
-func (run *Runner) Fetch(brandID, repairID string) (r []asserts.Assertion, err error) {
+// Fetch retrieves a stream with the repair with the given ids and any
+// auxiliary assertions. If revision>=0 Fetch sets in the request the
+// If-None-Match header with an E-Tag based on it, and
```
Comment (chipaca) [inline cmd/snap-repair/runner.go]: Hmm, that's rather awkward wording (in particular the “sets in the request” bit). May I suggest `If revision >= 0 create an ETag based on the given revision, and set the request's If-None-Match header to this ETag.`? (also note ETag doesn't have a hyphen)

**Pull request merged**